### ISSUE TITLE
Autoenv stuck in infinite loop

### ISSUE THREAD

**Issue by kdeldycke** [kennethreitz/autoenv#139]
Autoenv at 644903c works fine for me but latest versions produce what looks like an infinite loop, as my terminal is stuck when launched and does't give me a prompt.

I was waiting for #136 to be merged, unfortunately 3888c0b is still stuck in infinite loops.

This was reproduced on two machines running OSX 10.11.6 with [my custom `dotfiles`](https://github.com/kdeldycke/dotfiles).

I'm going to update this ticket with new findings as I continue my investigations.

**Comment by kdeldycke** [comment]
Pinpointed the issue as starting with commit 8a457ef46aa94afc1c811580816b510ee4764b58. Previous commit 879d11a2b68a4e9a802606ac6abe9ba8b5c755a3 works fine.

**Comment by dasJ** [comment]
Great, do your start your shell in a directory where any part is a symlink?

**Comment by kdeldycke** [comment]
Now that's funny:
```bash
$ stat -c '%m' \"$PWD\"
stat: illegal option -- c
usage: stat [-FlLnqrsx] [-f format] [-t timefmt] [file ...]
$
```

**Comment by kdeldycke** [comment]
This issue happens in a standard folder, where all parent folders are on the same volume, and none of them are symlinks.

**Comment by kdeldycke** [comment]
Just added a `set -x` to the top of `activate.sh`. Here the evidence of the infinite loop:

```bash
Last login: Thu Sep  8 16:49:52 on ttys010
++ AUTOENV_AUTH_FILE=/Users/kdeldycke/.autoenv_authorized
++ AUTOENV_ENV_FILENAME=.env
++ which gsha1sum
++ enable_autoenv
++ cd /Users/kdeldycke/dotfiles/dotfiles-common/.autoenv
++ autoenv_cd /Users/kdeldycke/dotfiles/dotfiles-common/.autoenv
++ command -v chdir
++ builtin cd /Users/kdeldycke/dotfiles/dotfiles-common/.autoenv
++ '[' 0 -eq 0 ']'
++ autoenv_init
++ local _mountpoint _files _orderedfiles _sedregexp
++ '[' 15 '!=' darwin15 ']'
+++ df -Th /Users/kdeldycke/dotfiles/dotfiles-common/.autoenv
+++ awk 'END{print $NF}'
++ _mountpoint=
++ _sedregexp=-E
+++ echo /Users/kdeldycke/dotfiles/dotfiles-common/.autoenv
+++ sed -E s:/+:/:g
++ PWD=/Users/kdeldycke/dotfiles/dotfiles-common/.autoenv
++ command -v chdir
++ builtin cd /Users/kdeldycke/dotfiles/dotfiles-common/.autoenv
+++ _hadone=
++++ pwd -P
+++ '[' /Users/kdeldycke/dotfiles/dotfiles-common/.autoenv '!=' '' ']'
++++ pwd -P
+++ _file=/Users/kdeldycke/dotfiles/dotfiles-common/.autoenv/.env
+++ '[' -f /Users/kdeldycke/dotfiles/dotfiles-common/.autoenv/.env ']'
+++ command -v chdir
++++ pwd -P
+++ builtin cd /Users/kdeldycke/dotfiles/dotfiles-common/.autoenv/..
++++ pwd -P
+++ '[' /Users/kdeldycke/dotfiles/dotfiles-common '!=' '' ']'
++++ pwd -P
+++ _file=/Users/kdeldycke/dotfiles/dotfiles-common/.env
+++ '[' -f /Users/kdeldycke/dotfiles/dotfiles-common/.env ']'
+++ command -v chdir
++++ pwd -P
+++ builtin cd /Users/kdeldycke/dotfiles/dotfiles-common/..
++++ pwd -P
+++ '[' /Users/kdeldycke/dotfiles '!=' '' ']'
++++ pwd -P
+++ _file=/Users/kdeldycke/dotfiles/.env
+++ '[' -f /Users/kdeldycke/dotfiles/.env ']'
+++ command -v chdir
++++ pwd -P
+++ builtin cd /Users/kdeldycke/dotfiles/..
++++ pwd -P
+++ '[' /Users/kdeldycke '!=' '' ']'
++++ pwd -P
+++ _file=/Users/kdeldycke/.env
+++ '[' -f /Users/kdeldycke/.env ']'
+++ command -v chdir
++++ pwd -P
+++ builtin cd /Users/kdeldycke/..
++++ pwd -P
+++ '[' /Users '!=' '' ']'
++++ pwd -P
+++ _file=/Users/.env
+++ '[' -f /Users/.env ']'
+++ command -v chdir
++++ pwd -P
+++ builtin cd /Users/..
++++ pwd -P
+++ '[' / '!=' '' ']'
++++ pwd -P
+++ _file=//.env
+++ '[' -f //.env ']'
+++ command -v chdir
++++ pwd -P
+++ builtin cd //..
++++ pwd -P
+++ '[' // '!=' '' ']'
++++ pwd -P
+++ _file=///.env
+++ '[' -f ///.env ']'
+++ command -v chdir
++++ pwd -P
+++ builtin cd ///..
++++ pwd -P
+++ '[' / '!=' '' ']'
++++ pwd -P
+++ _file=//.env
+++ '[' -f //.env ']'
+++ command -v chdir
++++ pwd -P
+++ builtin cd //..
++++ pwd -P
+++ '[' // '!=' '' ']'
++++ pwd -P
+++ _file=///.env
+++ '[' -f ///.env ']'
+++ command -v chdir
++++ pwd -P
+++ builtin cd ///..
++++ pwd -P
+++ '[' / '!=' '' ']'
++++ pwd -P
+++ _file=//.env
+++ '[' -f //.env ']'
+++ command -v chdir
++++ pwd -P
+++ builtin cd //..
++++ pwd -P
+++ '[' // '!=' '' ']'
++++ pwd -P
+++ _file=///.env
+++ '[' -f ///.env ']'
+++ command -v chdir
++++ pwd -P
+++ builtin cd ///..
++++ pwd -P
+++ '[' / '!=' '' ']'
++++ pwd -P
+++ _file=//.env
+++ '[' -f //.env ']'
+++ command -v chdir
++++ pwd -P
+++ builtin cd //..
++++ pwd -P
+++ '[' // '!=' '' ']'
++++ pwd -P
+++ _file=///.env
+++ '[' -f ///.env ']'
+++ command -v chdir
++++ pwd -P
+++ builtin cd ///..
++++ pwd -P
+++ '[' / '!=' '' ']'
++++ pwd -P
+++ _file=//.env
+++ '[' -f //.env ']'
+++ command -v chdir
++++ pwd -P
(...)
```

**Comment by kdeldycke** [comment]
Looks like an edge case at the terminal condition when backtracking up to the filesystem root.

**Comment by kdeldycke** [comment]
The `stat: illegal option -- c` mentioned above is unrelated and has been fixed in aece1ba3b557c1ca49b662e117323972d5227877.

**Comment by dasJ** [comment]
Yes, that's what I just wanted to write. For Darwin, it's:
```
_mountpoint="`df "${PWD}" | awk 'END{print $NF}'`"
```

**Comment by dasJ** [comment]
This seems to be the issue:
```
+++ df -Th /Users/kdeldycke/dotfiles/dotfiles-common/.autoenv
+++ awk 'END{print $NF}'
++ _mountpoint=
```

Can you give the output of the df without the awk?

**Comment by kdeldycke** [comment]
@dasJ Forget about that `stat`: it was just a side effect of trying the old 8a457ef commit. 😀

**Comment by kdeldycke** [comment]
You likely find the culprit:
```bash
$ df -Th /Users/kdeldycke/dotfiles/dotfiles-common/.autoenv
df: only one -T option may be specified
$
```

**Comment by dasJ** [comment]
Do you use the GNU coreutils or the BSD ones?

**Comment by kdeldycke** [comment]
Looks like I use the system default, aka BSD ones:
```bash
$ which df
/bin/df
$ df
$ gdf
Filesystem                                     1K-blocks      Used        Available Use% Mounted on
/dev/disk2                                    3033390096 491362236       2541771860  17% /
(...)
$
```

**Comment by dasJ** [comment]
Does it work for you with the current master?

**Comment by dirk-olmes** [comment]
I'm stumbling over the same issue - albeit on CentOS 6, not on OSX

I can bisect the error down to the same commit that @kdeldycke mentions in the second comment above.

**Comment by ace13** [comment]
Ran into this same issue as well, CentOS 6.8

I think this might have something to do with it;
```
$ stat -c %m /home/ace/project
?
```

Here's a log with `set -x`
```
++ AUTOENV_AUTH_FILE=/home/ace/.autoenv_authorized
++ AUTOENV_ENV_FILENAME=.env
++ /usr/bin/which --tty-only --read-alias --show-dot --show-tilde gsha1sum
++ alias
++ /usr/bin/which --tty-only --read-alias --show-dot --show-tilde sha1sum
++ alias
++ enable_autoenv
++ cd /home/ace/project
++ autoenv_cd /home/ace/project
++ command -v chdir
++ builtin cd /home/ace/project
++ '[' 0 -eq 0 ']'
++ autoenv_init
++ local _mountpoint _files _orderedfiles _sedregexp
++ '[' linux-gnu '!=' linux-gnu ']'
++++ pwd -P
+++ stat -c %m /home/ace/project
++ _mountpoint='?'
++ _sedregexp=-r
+++ echo /home/ace/project
+++ sed -r s:/+:/:g
++ PWD=/home/ace/project
++ command -v chdir
++ builtin cd /home/ace/project
+++ _hadone=
++++ pwd -P
+++ '[' /home/ace/project '!=' '?' ']'
++++ pwd -P
+++ _file=/home/ace/project/.env
+++ '[' -f /home/ace/project/.env ']'
+++ '[' -z '' ']'
+++ echo -n /home/ace/project/.env
+++ _hadone=1
+++ command -v chdir
++++ pwd -P
+++ builtin cd /home/ace/project/..
++++ pwd -P
+++ '[' /home/ace '!=' '?' ']'
++++ pwd -P
+++ _file=/home/ace/.env
+++ '[' -f /home/ace/.env ']'
+++ '[' -z 1 ']'
+++ echo -n '
/home/ace/.env'
+++ command -v chdir
++++ pwd -P
+++ builtin cd /home/ace/..
++++ pwd -P
+++ '[' /home '!=' '?' ']'
++++ pwd -P
+++ _file=/home/.env
+++ '[' -f /home/.env ']'
+++ command -v chdir
++++ pwd -P
+++ builtin cd /home/..
++++ pwd -P
+++ '[' / '!=' '?' ']'
++++ pwd -P
+++ _file=//.env
+++ '[' -f //.env ']'
+++ command -v chdir
++++ pwd -P
+++ builtin cd //..
++++ pwd -P
+++ '[' // '!=' '?' ']'
++++ pwd -P
+++ _file=///.env
+++ '[' -f ///.env ']'
+++ command -v chdir
++++ pwd -P
+++ builtin cd ///..
++++ pwd -P
+++ '[' / '!=' '?' ']'
++++ pwd -P
+++ _file=//.env
+++ '[' -f //.env ']'
+++ command -v chdir
++++ pwd -P
+++ builtin cd //..
++++ pwd -P
+++ '[' // '!=' '?' ']'
++++ pwd -P
+++ _file=///.env
+++ '[' -f ///.env ']'
+++ command -v chdir
++++ pwd -P
+++ builtin cd ///..
++++ pwd -P
+++ '[' / '!=' '?' ']'
```

**Comment by dasJ** [comment]
Does #142 (or the master of my fork) fix the issue?
I can confirm it works on Linux with the most recent shell version

**Comment by kdeldycke** [comment]
Just fixed my issue with commit https://github.com/kdeldycke/dotfiles/commit/ac56d7e859f1a5f389c1ecf91c6598c8be4dced8 . Top of the `master` branch is working fine in the end. The source of my issue was a local `alias` of `df`. I'll then close this issue.

**Comment by dirk-olmes** [comment]
@kdeldycke I understand that your issue was caused by a local alias to df. My issue on CentOS is something different, though. I just double-checked that I do not have an alias to df (I don't) and the latest trunk still shows the issue.

The fix proposed by @dasJ however fixes the issue on CentOS for me, too.

**Comment by kdeldycke** [comment]
Do you want me to re-open this issue then?

**Comment by dirk-olmes** [comment]
I'd say leave it closed - you were able to resolve your issue. The fix for my issue is in #142
